### DISCUSSION TITLE
The material can't be defined on the boundary in 1D mesh

### DISCUSSION THREAD

**Discussion by xuxiaobei1995** [idaholab/moose#22162] [Q&A General] [answered]
Hi all,

Recently I was trying using 1D mesh in MOOSE. In the test case I defined a material on a boundary but to cause an error "the material object contains the following boundary ids that do not exist on the mesh". However, the boundary can be recognized in the BC. The 1D mesh file is generated using Gmsh. The input file is:

```
[Mesh]
  type = FileMesh
  file = 1d_mesh_test.msh
[]

[Variables]
  [./u]
  [../]
[]

[Kernels]
  [./diff]
    type = Diffusion
    variable = u
    block = '11 12'
  [../]
[]

[BCs]
  [./left_left]
    type = DirichletBC
    variable = u
    boundary = 7
    value = 1
  [../]
  [./left_right]
    type = DirichletBC
    variable = u
    boundary = 8
    value = 0
  [../]
  [./right_left]
    type = DirichletBC
    variable = u
    boundary = 9
    value = 1
  [../]
  [./right_right]
    type = DirichletBC
    variable = u
    boundary = 10
    value = 0
  [../]
[]

[Materials]
  [mat]
    type = GenericConstantMaterial
    prop_names = 'm'
    prop_values = 1.0
    boundary = 8
  []
[]

[Executioner]
  type = Steady
  solve_type = 'PJFNK'
  petsc_options_iname = '-pc_type -pc_hypre_type'
  petsc_options_value = 'hypre boomeramg'
[]

[Outputs]
  exodus = true
[]
```

The Gmsh geometry file is:

```
Point(1) = {0, 0.0, 0};
Point(2) = {0.0026, 0.0, 0};
Point(3) = {0.0027, 0, 0};
Point(4) = {0.003, 0, 0};
Line(5) = {1, 2};
Line(6) = {3, 4};
Transfinite Line {5} = 21 Using Progression 1;
Transfinite Line {6} = 5 Using Progression 1;
Physical Point(7) = {1};
Physical Point(8) = {2};
Physical Point(9) = {3};
Physical Point(10) = {4};
Physical Line(11) = {5};
Physical Line(12) = {6};
```

The Gmsh mesh file is:

```
$MeshFormat
2 0 8
$EndMeshFormat
$Nodes
26
1 0 0 0
2 0.0026 0 0
3 0.0027 0 0
4 0.003 0 0
5 0.0001299999999998189 0 0
6 0.0002599999999996378 0 0
7 0.0003899999999994567 0 0
8 0.0005199999999992756 0 0
9 0.0006499999999990944 0 0
10 0.0007799999999989742 0 0
11 0.0009099999999988538 0 0
12 0.001039999999998733 0 0
13 0.001169999999998613 0 0
14 0.001299999999998493 0 0
15 0.001429999999998372 0 0
16 0.001559999999998252 0 0
17 0.001689999999998132 0 0
18 0.001819999999998012 0 0
19 0.001949999999997891 0 0
20 0.002079999999998313 0 0
21 0.002209999999998735 0 0
22 0.002339999999999157 0 0
23 0.002469999999999578 0 0
24 0.002775 0 0
25 0.00285 0 0
26 0.002925 0 0
$EndNodes
$Elements
28
1 15 3 7 1 0 1
2 15 3 8 2 0 2
3 15 3 9 3 0 3
4 15 3 10 4 0 4
5 1 3 11 5 0 1 5
6 1 3 11 5 0 5 6
7 1 3 11 5 0 6 7
8 1 3 11 5 0 7 8
9 1 3 11 5 0 8 9
10 1 3 11 5 0 9 10
11 1 3 11 5 0 10 11
12 1 3 11 5 0 11 12
13 1 3 11 5 0 12 13
14 1 3 11 5 0 13 14
15 1 3 11 5 0 14 15
16 1 3 11 5 0 15 16
17 1 3 11 5 0 16 17
18 1 3 11 5 0 17 18
19 1 3 11 5 0 18 19
20 1 3 11 5 0 19 20
21 1 3 11 5 0 20 21
22 1 3 11 5 0 21 22
23 1 3 11 5 0 22 23
24 1 3 11 5 0 23 2
25 1 3 12 6 0 3 24
26 1 3 12 6 0 24 25
27 1 3 12 6 0 25 26
28 1 3 12 6 0 26 4
$EndElements
```

Does anyone know how to solve this issue? Thanks.

Best Regards,

Xiaobei

**Comment by GiudGiud** [comment, marked as answer]
Hello

You don’t really need to boundary restrict this material. Looking at your input I don’t think it changes anything to define it over the whole mesh. 

Guillaume

**Reply by xuxiaobei1995** [reply]
Hi GiudGiud,

Actually this is just a test case. What I really need is to apply the thermal contact in HeatConduction module to this 1D mesh. The thermal contact restricts the materials to the gap boundaries so I have to solve this issue.

**Reply by GiudGiud** [reply]
Open the Mesh in paraview (you can use —mesh-only and exodus output in moose for that)
And look at the id or name of the side set
Looking at the gmsh file I m not sure it exists?

**Reply by xuxiaobei1995** [reply]
I opened the output file in Paraview:

![image](https://user-images.githubusercontent.com/34068460/191785013-6d5d1326-f8bd-4c31-92af-50665b821bbe.png)

it seems that the boundaries of the 1D mesh can only be recognized as nodesets but not sidesets. It seems reasonable because the boundary only has one node in 1D mesh.

I also read the code of BoundaryRestrictable, on which the Material is based:

![image](https://user-images.githubusercontent.com/34068460/191791325-a89d68ee-9f38-471a-8acd-729cb6f7b13f.png)

The _bnd_nodal in Material is false so the Material can not be restricted to nodesets. Maybe I could set the _bnd_nodal in Material to be true to solve this issue?

**Reply by GiudGiud** [reply]
I can see that. 

Is your target case 1D as well?

**Reply by xuxiaobei1995** [reply]
You mean setting the dim = 1 in [Mesh]? I did this but the error still existed ...

**Reply by grmnptr** [reply]
What version of paraview  are you using? Are you using the legacy exodus reader?

**Reply by grmnptr** [reply]
Also, can you try to name your physical points/lines?
```
Physical Point("left_begin") = {1};
Physical Point("left_end") = {2};
Physical Point("right_begin") = {3};
Physical Point("right_end") = {4};
Physical Line("left_domain") = {5};
Physical Line("right_domain") = {6};
```

**Reply by xuxiaobei1995** [reply]
> What version of paraview are you using? Are you using the legacy exodus reader?

I use Paraview 4.4.0. 



> Also, can you try to name your physical points/lines?
> 
> ```
> Physical Point("left_begin") = {1};
> Physical Point("left_end") = {2};
> Physical Point("right_begin") = {3};
> Physical Point("right_end") = {4};
> Physical Line("left_domain") = {5};
> Physical Line("right_domain") = {6};
> ```

I tried but Gmsh raised error. Maybe it is because my Gmsh version is so old (2.2.6) that only ID is supported.

**Reply by GiudGiud** [reply]
Paraview 4.4 is definitely very old, they are at 11 now. 
If you upgrade to 11 or 10, you will have to make sure to select the Legacy Exodus reader

**Reply by GiudGiud** [reply]
Is your final case (the one you are making this test case for) also 1D ?

**Reply by xuxiaobei1995** [reply]
> Is your final case (the one you are making this test case for) also 1D ?

Yes, it is a 1D case.

**Reply by GiudGiud** [reply]
Have you tried Peter's idea?
Any luck with modifying the boundary material code?

**Reply by xuxiaobei1995** [reply]
> Have you tried Peter's idea? 

I tried adding name to the boundaries but Gmsh raised error. Maybe it is because the version of my Gmsh only supports IDs.

> Any luck with modifying the boundary material code?

I have not done it because recently I'm busy with other things. I will try it after finishing those things (maybe several days later).
